File: 剑指offer/59.1.滑动窗口的最大值.py
from typing import List
from collections import deque


class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        left = 0
        right = 0

        window = deque()
        ans = []
        while right < len(nums):
            while window and window[-1] < nums[right]:
                window.pop()

            window.append(nums[right])
            right += 1

            if right - left < k:
                continue
            elif right - left > k:
                if nums[left] == window[0]:
                    window.popleft()
                left += 1

            ans.append(window[0])

        return ans

File: 剑指offer/59.1.滑动窗口的最大值.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k)
    {
        deque<int> q;
        int n = nums.size();

        if (n == 0 || k == 0) {
            return {};
        }

        for (int i = 0; i < k; i++) {
            while (!q.empty() && nums[q.back()] < nums[i]) {
                q.pop_back();
            }
            q.push_back(i);
        }

        vector<int> ans;
        ans.push_back(nums[q.front()]);

        for (int i = k; i < n; i++) {
            while (!q.empty() && nums[q.back()] < nums[i]) {
                q.pop_back();
            }
            q.push_back(i);
            if (i - k == q.front()) {
                q.pop_front();
            }
            ans.push_back(nums[q.front()]);
        }

        return ans;
    }
};